File: day1/trebuchet.go
package main

import (
	"bufio"
	"log"
	"os"
	"strconv"
	"strings"
	"unicode"
)

// https://adventofcode.com/2023/day/1#part2

func main() {
	lines := readInput("input.txt")
	myDict := initDict()
	total := 0

	for _, line := range lines {
		first := extractFirst(line, myDict)
		last := extractLast(line, myDict)
		sum, _ := strconv.Atoi(first + last)
		total += sum
	}

	log.Println(total)
}

func readInput(path string) []string {
	f, err := os.Open(path)
	if err != nil {
		log.Fatalf("unable to read file: %v", err)
	}

	defer f.Close()

	var lines []string
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	return lines
}

func initDict() map[string]string {
	myDict := make(map[string]string)

	myDict["one"] = "1"
	myDict["two"] = "2"
	myDict["three"] = "3"
	myDict["four"] = "4"
	myDict["five"] = "5"
	myDict["six"] = "6"
	myDict["seven"] = "7"
	myDict["eight"] = "8"
	myDict["nine"] = "9"

	return myDict
}

func extractFirst(line string, myDict map[string]string) string {
	for i := 0; i < len(line); i++ {
		if unicode.IsDigit(rune(line[i])) {
			return string(line[i])
		} else {
			for k, v := range myDict {
				if strings.Contains(line[:i+1], k) {
					return v
				}
			}
		}
	}
	return ""
}

func extractLast(line string, myDict map[string]string) string {
	for i := len(line) - 1; i >= 0; i-- {
		if unicode.IsDigit(rune(line[i])) {
			return string(line[i])
		} else {
			for k, v := range myDict {
				if strings.Contains(line[i:], k) {
					return v
				}
			}
		}
	}
	return ""
}

File: day1/python/trebuchet.py
"""Advent of code day1 solution"""


def read_input(path: str) -> list[str]:
    """Read lines from input file.

    Keyword arguments:
    path -- path to input.* file
    """
    ll = []
    with open(path, "r", encoding="utf-8") as infile:
        for entry in infile:
            ll.append(entry[:-1])
    return ll


def init_dict() -> dict:
    """Initiates value dictionary"""
    values = {
        "one": "1",
        "two": "2",
        "three": "3",
        "four": "4",
        "five": "5",
        "six": "6",
        "seven": "7",
        "eight": "8",
        "nine": "9",
    }
    return values


def extract_first(entry: str, vd: dict) -> str:
    """Extracts the first digit or string value in a line

    Keyword arguments:
    entry -- line from input.* file
    vd -- value dictionary
    """
    for i, c in enumerate(entry):
        if c.isdigit():
            return c
        for k, v in vd.items():
            if k in entry[: i + 1]:
                return v
    return ""


def extract_last(entry: str, vd: dict) -> str:
    """Extracts the last digit or string value in a line

    Keyword arguments:
    entry -- line from input.* file
    vd -- value dictionary
    """
    for i, c in enumerate(reversed(entry)):
        if c.isdigit():
            return c
        for k, v in vd.items():
            if k in entry[len(entry) - i - 1 :]:
                return v
    return ""


if __name__ == "__main__":
    lines = read_input("../input.txt")
    value_dict = init_dict()

    TOTAL_SUM = 0
    for line in lines:
        first = extract_first(line, value_dict)
        last = extract_last(line, value_dict)
        num_sum = int(first + last)
        print(f"line: {line}, first: {first}, last: {last}")
        TOTAL_SUM += num_sum
    print(f"Total sum: {TOTAL_SUM}")
